posts/gups/gups.cu

```diff
@@ -0,0 +1,598 @@
+
+  uint64_t ran = 2;
+  while (i > 0) {
+    uint64_t temp = 0;
+    for (int j = 0; j < 64; j++) {
+      if ((ran >> j) & 1) {
+        temp ^= c_m2[j];
+      }
+    }
+    ran = temp;
+    i -= 1;
+    if ((n >> i) & 1) {
+      ran = (ran << 1) ^ ((int64_t)ran < 0 ? POLY : 0);
+    }
+  }
+
+  return ran;
+}
+
+enum test_t
+{
+  UPDATE = 0,
+  READ,
+  WRITE,
+  READ_WRITE,
+  UPDATE_NO_LOOP,
+};
+
+const char* test_name[] = {
+    "Gupdates/s ATOM.CAS.64(loop)",
+    "Greads/s LD.64",
+    "Gwrites/s STG.64",
+    "Greads+writes/s LDG.64+STG.64",
+    "Gupdates/s ATOM.CAS.64(no_loop)",
+};
+
+template <test_t T_TYPE>
+__global__ void
+d_bench(size_t n, size_t working_set, benchtype* t, int accesses_per_elem)
+{
+  size_t num_threads = gridDim.x * blockDim.x;
+  size_t thread_num = blockIdx.x * blockDim.x + threadIdx.x;
+  size_t start = thread_num * accesses_per_elem * n / num_threads;
+  size_t end = (thread_num + 1) * accesses_per_elem * n / num_threads;
+  benchtype ran;
+  ran.data = d_starts(start);
+
+#pragma unroll
+  for (ptrdiff_t i = start; i < end; ++i) {
+    ran.data = (ran.data << 1) ^ ((int64_t)ran.data < 0 ? POLY : 0);
+    unsigned long long int *address, old, assumed;
+    address = (unsigned long long int*)&t[ran.data & (working_set - 1)].data;
+    switch (T_TYPE) {
+    case READ:
+      old = *address;
+      if (old == n) { // basically never executes
+        *address = n + 1;
+      }
+      break;
+    case WRITE:
+      *address = 1;
+      break;
+    case READ_WRITE:
+      *address += 1;
+      break;
+    case UPDATE_NO_LOOP:
+      old = *address;
+      assumed = old;
+      old = atomicCAS(address, assumed, assumed ^ ran.data);
+      break;
+    case UPDATE:
+      old = *address;
+      do {
+        assumed = old;
+        old = atomicCAS(address, assumed, assumed ^ ran.data);
+      } while (assumed != old);
+      break;
+    }
+  }
+}
+
+template <test_t T_TYPE>
+__global__ void d_bench_shmem(size_t n_shmem, int accesses_per_elem_sh)
+{
+  extern __shared__ benchtype extern_mem[];
+  benchtype* t = (benchtype*)&extern_mem;
+
+  size_t num_threads = gridDim.x * blockDim.x;
+  size_t thread_num = blockIdx.x * blockDim.x + threadIdx.x;
+#ifdef STATIC_SHMEM
+  // ignore any compiler warnings about not using n_shmem!
+  const size_t n_shmem_l = NSHMEM;
+#else
+  const size_t n_shmem_l = n_shmem;
+#endif
+  size_t start = thread_num * accesses_per_elem_sh * n_shmem_l / num_threads;
+  size_t end = (thread_num + 1) * accesses_per_elem_sh * n_shmem_l / num_threads;
+  benchtype ran;
+  ran.data = d_starts(start);
+
+#pragma unroll
+  for (ptrdiff_t i = start; i < end; ++i) {
+    ran.data = (ran.data << 1) ^ ((int64_t)ran.data < 0 ? POLY : 0);
+    unsigned long long int *address, old, assumed;
+    address = (unsigned long long int*)&t[ran.data % n_shmem_l].data;
+    switch (T_TYPE) {
+    case READ:
+      old = *address;
+      if (old == n_shmem_l) { // basically never executes
```

Comment:
Could be a candidate for `__builtin_expect`: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#builtin-expect

Reply:
Thanks for the info. `__builtin_expect` doesn't improve performance for global memory GUPS. But it could help with shared memory GUPS. Performance benchmarking is WIP.

Reply:
Tested it for the shmem as well, no impact.
`__builtin_expect()` is described as a compiler hint, but it's different from preprocessor type compiler hints (e.g. `#pragma unroll`) and having it is not guaranteed to be harmless.
I think we should keep the code as is since it could end up harming the behavior.